File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/item_detail/bodies/components/GoingStrip.kt
@file:Suppress("PackageNaming", "MagicNumber", "FunctionNaming", "LongMethod")

package app.pantopus.android.ui.screens.mailbox.item_detail.bodies.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.data.api.models.mailbox.v2.PartyAttendee
import app.pantopus.android.data.api.models.mailbox.v2.PartyDetailDto
import app.pantopus.android.data.api.models.mailbox.v2.PartyRsvpStatus
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * A17.9 — "+N going" friend pile. Section header carries the headcount
 * + maybe count + a "See all" affordance; body wraps the friend avatars
 * with a +N plus-one badge stacked on each. In the going state a
 * primary-tinted "You" avatar is prepended so the user reads as part of
 * the pile.
 */
@Composable
fun GoingStrip(
    party: PartyDetailDto,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .testTag("partyGoingStrip"),
    ) {
        Header(party)
        HorizontalDivider(color = PantopusColors.appBorderSubtle)
        AvatarFlow(party)
    }
}

@Composable
private fun Header(party: PartyDetailDto) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                text = "${party.headcount} GOING · ${party.maybeCount} MAYBE",
                fontSize = 11.sp,
                fontWeight = FontWeight.Black,
                letterSpacing = 0.5.sp,
                color = PantopusColors.appTextSecondary,
            )
            Text(
                text =
                    if (party.rsvp == PartyRsvpStatus.Going) {
                        "Including you + ${party.plusOneCount}"
                    } else {
                        "5 friends · 2 plus-ones"
                    },
                fontSize = 11.sp,
                color = PantopusColors.appTextSecondary,
            )
        }
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                text = "See all",
                fontSize = 11.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.primary600,
            )
            PantopusIconImage(
                icon = PantopusIcon.ChevronRight,
                contentDescription = null,
                size = 12.dp,
                tint = PantopusColors.primary600,
            )
        }
    }
}

@Composable
private fun AvatarFlow(party: PartyDetailDto) {
    FlowRow(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
        spacing = Spacing.s2,
    ) {
        if (party.rsvp == PartyRsvpStatus.Going) {
            YouAvatar(plusCount = party.plusOneCount)
        }
        party.goingAttendees.forEach { attendee ->
            AttendeeAvatar(attendee = attendee)
        }
    }
}

@Composable
private fun AttendeeAvatar(attendee: PartyAttendee) {
    Column(
        modifier =
            Modifier
                .widthIn(min = 44.dp)
                .testTag("partyGoingStrip_attendee_${attendee.id}"),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        Box(modifier = Modifier.size(40.dp)) {
            Box(
                modifier =
                    Modifier
                        .size(40.dp)
                        .clip(CircleShape)
                        .background(attendee.accent.color()),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = attendee.initials,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.appTextInverse,
                )
            }
            if (attendee.plusCount > 0) {
                PlusBadge(text = "+${attendee.plusCount}", modifier = Modifier.align(Alignment.BottomEnd))
            }
        }
        Text(
            text = attendee.name,
            fontSize = 9.5.sp,
            fontWeight = FontWeight.SemiBold,
            color = PantopusColors.appTextSecondary,
            maxLines = 1,
        )
    }
}

@Composable
private fun YouAvatar(plusCount: Int) {
    Column(
        modifier =
            Modifier
                .widthIn(min = 44.dp)
                .testTag("partyGoingStrip_you"),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        Box(modifier = Modifier.size(40.dp)) {
            Box(
                modifier =
                    Modifier
                        .size(40.dp)
                        .clip(CircleShape)
                        .background(
                            Brush.linearGradient(
                                listOf(PantopusColors.primary500, PantopusColors.primary700),
                            ),
                        )
                        .border(2.5.dp, PantopusColors.primary300, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = "You",
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.appTextInverse,
                )
            }
            PlusBadge(
                text = if (plusCount > 0) "+$plusCount" else "+1",
                modifier = Modifier.align(Alignment.BottomEnd),
            )
        }
        Text(
            text = "You",
            fontSize = 9.5.sp,
            fontWeight = FontWeight.Black,
            color = PantopusColors.primary700,
            maxLines = 1,
        )
    }
}

@Composable
private fun PlusBadge(
    text: String,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier =
            modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.categoryParty)
                .border(2.dp, PantopusColors.appSurface, RoundedCornerShape(Radii.pill))
                .padding(horizontal = Spacing.s1, vertical = 1.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = text,
            fontSize = 9.sp,
            fontWeight = FontWeight.Black,
            color = PantopusColors.appTextInverse,
        )
    }
}

/** Map the data-side accent enum onto the design system. View stays out
 *  of the hex palette. */
private fun PartyAttendee.AccentTint.color(): Color =
    when (this) {
        PartyAttendee.AccentTint.Home -> PantopusColors.home
        PartyAttendee.AccentTint.Personal -> PantopusColors.personal
        PartyAttendee.AccentTint.Business -> PantopusColors.business
        PartyAttendee.AccentTint.Warning -> PantopusColors.handyman
        PartyAttendee.AccentTint.Error -> PantopusColors.error
        PartyAttendee.AccentTint.Primary -> PantopusColors.primary600
        PartyAttendee.AccentTint.Party -> PantopusColors.categoryParty
    }

/** Minimal flow layout — wraps children left-to-right and starts a new
 *  line when the next item would overflow the available width. */
@Composable
private fun FlowRow(
    modifier: Modifier = Modifier,
    spacing: Dp,
    content: @Composable () -> Unit,
) {
    Layout(modifier = modifier, content = content) { measurables, constraints ->
        val spacingPx = spacing.roundToPx()
        var x = 0
        var y = 0
        var rowHeight = 0
        val placements = mutableListOf<Triple<androidx.compose.ui.layout.Placeable, Int, Int>>()
        measurables.forEach { measurable ->
            val placeable = measurable.measure(constraints.copy(minWidth = 0))
            if (x + placeable.width > constraints.maxWidth && x > 0) {
                x = 0
                y += rowHeight + spacingPx
                rowHeight = 0
            }
            placements += Triple(placeable, x, y)
            x += placeable.width + spacingPx
            rowHeight = maxOf(rowHeight, placeable.height)
        }
        val totalHeight = y + rowHeight
        layout(constraints.maxWidth, totalHeight) {
            placements.forEach { (placeable, px, py) -> placeable.placeRelative(px, py) }
        }
    }
}

File: frontend/apps/ios/Pantopus/Features/Mailbox/MailDetail/Variants/Components/GoingStrip.swift
//
//  GoingStrip.swift
//  Pantopus
//
//  A17.9 — "+N going" friend pile. Section header carries the headcount
//  + maybe count + a "See all" affordance; body wraps the friend avatars
//  with a +N plus-one badge stacked on each. In the going state a
//  primary-tinted "You" avatar is prepended so the user reads as part
//  of the pile alongside their friends.
//

import SwiftUI

@MainActor
struct GoingStrip: View {
    let party: PartyDetailDTO

    var body: some View {
        VStack(spacing: Spacing.s0) {
            header
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
            avatarFlow
        }
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        .accessibilityIdentifier("partyGoingStrip")
    }

    // MARK: - Header

    private var header: some View {
        HStack(spacing: Spacing.s2) {
            VStack(alignment: .leading, spacing: 2) {
                Text("\(party.headcount) GOING · \(party.maybeCount) MAYBE")
                    .font(.system(size: 11, weight: .heavy))
                    .tracking(0.5)
                    .foregroundStyle(Theme.Color.appTextSecondary)
                Text(party.rsvp == .going
                    ? "Including you + \(party.plusOneCount)"
                    : "5 friends · 2 plus-ones"
                )
                .font(.system(size: 11))
                .foregroundStyle(Theme.Color.appTextSecondary)
            }
            Spacer()
            HStack(spacing: 2) {
                Text("See all")
                    .font(.system(size: 11, weight: .bold))
                Icon(.chevronRight, size: 12, color: Theme.Color.primary600)
            }
            .foregroundStyle(Theme.Color.primary600)
        }
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s2)
    }

    // MARK: - Avatar pile

    private var avatarFlow: some View {
        FlowLayout(spacing: Spacing.s2) {
            if party.rsvp == .going {
                YouAvatar(plusCount: party.plusOneCount)
            }
            ForEach(party.goingAttendees) { attendee in
                AttendeeAvatar(attendee: attendee)
            }
        }
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s2)
    }
}

// MARK: - Avatars

private struct AttendeeAvatar: View {
    let attendee: PartyAttendee

    var body: some View {
        VStack(spacing: Spacing.s1) {
            ZStack(alignment: .bottomTrailing) {
                Text(attendee.initials)
                    .font(.system(size: 12, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
                    .frame(width: 40, height: 40)
                    .background(attendee.accent.swiftUIColor)
                    .clipShape(Circle())
                if attendee.plusCount > 0 {
                    PlusBadge(text: "+\(attendee.plusCount)")
                }
            }
            Text(attendee.name)
                .font(.system(size: 9.5, weight: .semibold))
                .foregroundStyle(Theme.Color.appTextSecondary)
                .lineLimit(1)
        }
        .frame(width: 44)
        .accessibilityIdentifier("partyGoingStrip_attendee_\(attendee.id)")
    }
}

private struct YouAvatar: View {
    let plusCount: Int

    var body: some View {
        VStack(spacing: Spacing.s1) {
            ZStack(alignment: .bottomTrailing) {
                Text("You")
                    .font(.system(size: 11, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
                    .frame(width: 40, height: 40)
                    .background(
                        LinearGradient(
                            colors: [Theme.Color.primary500, Theme.Color.primary700],
                            startPoint: .topLeading,
                            endPoint: .bottomTrailing
                        )
                    )
                    .clipShape(Circle())
                    .overlay(
                        Circle().stroke(Theme.Color.primary300, lineWidth: 2.5)
                    )
                PlusBadge(text: plusCount > 0 ? "+\(plusCount)" : "+1")
            }
            Text("You")
                .font(.system(size: 9.5, weight: .heavy))
                .foregroundStyle(Theme.Color.primary700)
                .lineLimit(1)
        }
        .frame(width: 44)
        .accessibilityIdentifier("partyGoingStrip_you")
    }
}

private struct PlusBadge: View {
    let text: String

    var body: some View {
        Text(text)
            .font(.system(size: 9, weight: .heavy))
            .foregroundStyle(Theme.Color.appTextInverse)
            .padding(.horizontal, Spacing.s1)
            .frame(minWidth: 16, minHeight: 16)
            .background(Theme.Color.categoryParty)
            .overlay(Capsule().stroke(Theme.Color.appSurface, lineWidth: 2))
            .clipShape(Capsule())
            .offset(x: 4, y: 4)
    }
}

extension PartyAttendee.AccentTint {
    /// Maps the data-side accent enum onto the design system. Keeps the
    /// view free of any raw hex while staying faithful to the design's
    /// gradient palette.
    var swiftUIColor: Color {
        switch self {
        case .home: Theme.Color.home
        case .personal: Theme.Color.personal
        case .business: Theme.Color.business
        case .warning: Theme.Color.handyman
        case .error: Theme.Color.error
        case .primary: Theme.Color.primary600
        case .party: Theme.Color.categoryParty
        }
    }
}

// MARK: - Flow layout for the wrapping avatar pile

/// Minimal flow layout — wraps children left-to-right and starts a new
/// line when the next item would overflow the available width.
private struct FlowLayout: Layout {
    let spacing: CGFloat

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache _: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        var x: CGFloat = 0
        var y: CGFloat = 0
        var rowHeight: CGFloat = 0
        var maxRowWidth: CGFloat = 0
        for subview in subviews {
            let size = subview.sizeThatFits(.unspecified)
            if x + size.width > maxWidth, x > 0 {
                maxRowWidth = max(maxRowWidth, x - spacing)
                x = 0
                y += rowHeight + spacing
                rowHeight = 0
            }
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
        }
        maxRowWidth = max(maxRowWidth, x - spacing)
        return CGSize(width: maxRowWidth, height: y + rowHeight)
    }

    func placeSubviews(
        in bounds: CGRect,
        proposal: ProposedViewSize,
        subviews: Subviews,
        cache _: inout ()
    ) {
        let maxWidth = proposal.width ?? bounds.width
        var x: CGFloat = bounds.minX
        var y: CGFloat = bounds.minY
        var rowHeight: CGFloat = 0
        for subview in subviews {
            let size = subview.sizeThatFits(.unspecified)
            if x + size.width > bounds.minX + maxWidth, x > bounds.minX {
                x = bounds.minX
                y += rowHeight + spacing
                rowHeight = 0
            }
            subview.place(at: CGPoint(x: x, y: y), proposal: .unspecified)
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
        }
    }
}

#Preview("Open") {
    GoingStrip(party: MailItemSampleData.partyInvite)
        .padding(Spacing.s4)
        .background(Theme.Color.appBg)
}

#Preview("Going") {
    GoingStrip(party: MailItemSampleData.partyInviteGoing)
        .padding(Spacing.s4)
        .background(Theme.Color.appBg)
}
